### DISCUSSION TITLE
mogrify removes tiff compression

### DISCUSSION THREAD

**Discussion by Jmuccigr** [ImageMagick/ImageMagick#2452] [Help]
I have a simple script I use to alter image resolution so that the final product fits on a standard A4 or US letter page. It does some math based on the size of the image, then uses `mogrify` to set the density appropriately. I noticed that when the file in question is a tiff with Group4 compression (common for scanned texts), the operation removes the compression (`None`). Obviously this can significantly increase the file size. 

A little testing shows that this happens with any compressed tiff (zip, lzw, or Group4), but not with png files where the compression stays at `Zip` or with jpegs where the compression level is unchanged.

I can prevent this by using `-define tiff:preserve-compression=true` in the command, but shouldn't `mogrify` just leave the compression alone, as it does with the other file types? That, at least, was the behavior I expected.

Here's some data from a few runs where I took some compressed tiffs and just used mogrify to reset the density to 100 ppi. Names are based the compression in use; the last one used the flags to preserve the compression

```
>identify -format "%f %C %B\n" $i;done
ccitt.tiff Group4 43512
rerez_ccitt.tiff None 52910
zip.tiff Zip 275760
rerez_zip.tiff None 1262732
lzw.tiff LZW 329944
rerez_lzw.tiff None 1262732
fax.tiff Fax 40194
rerez_fax.tiff None 52910
preserve_lzw.tiff LZW 329944 
```

**Comment by snibgo** [comment]
What version of IM, on what platform?

IM's default behaviour on TIFF compression seems to have changed recently.

**Reply by Jmuccigr** [reply]
> What version of IM, on what platform?

MacOS 10.13.6

Version: ImageMagick 7.0.10-28 Q16 x86_64 2020-08-18 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(3.1)
Delegates (built-in): bzlib fftw fontconfig freetype heic jng jp2 jpeg lcms ltdl lzma png tiff webp wmf x xml zlib

> IM's default behaviour on TIFF compression seems to have changed recently.

Yeah, I don't seem to remember this happening before, though I don't always use tiffs.

**Comment by fmw42** [comment]
My recollection is that it has always converted compressed tiffs to uncompressed tiffs, unless a compression was specified.

**Reply by Jmuccigr** [reply]
> My recollection is that it has always converted compressed tiffs to uncompressed tiffs, unless a compression was specified.

You may be right. I expected though, that for consistency, it would leave the compression as it finds it.

**Reply by fmw42** [reply]
I think it might have been the way libtiff handled things in the past. That is why the define was created. Or you could always use -compress in your commands to set it to whatever you want.

**Reply by Jmuccigr** [reply]
I can fix the script; my question really is whether this **should** be the behavior of mogrify. I’d say no. The user shouldn’t have to worry about changes to the file other than those requested, as is the case with the other formats.

Happy to make this a feature request.

**Comment by snibgo** [comment]
> I expected though, that for consistency, it would leave the compression as it finds it.

A long time ago (not _recently_, my mistake), that's how IM TIFF compression worked. But users complained.

The problem was that some TIFF compression methods are lossy. (Unlike PNG, which is always lossless.) Read a TIFF image, do some processing such as adding colour, then write it to a TIFF file. But the processing looks wrong, eg the added colour has vanished. What has happened? It turned out that the input file had lossy compression, eg Group4, and that was used for the output.

So IM changed, to not carry input TIFF compression to output. True, Zip and LZW compressions are lossless, so those methods could have been carried forwards. But I suppose it was simpler to have a blanket rule.

**Reply by Jmuccigr** [reply]
Would this logic not also apply in the case of jpeg?

**Reply by fmw42** [reply]
Jpeg is always lossy. So there is no choice of non-lossy option. So ImageMagick tries to preserve the input compression quality.

**Reply by Jmuccigr** [reply]
Right but my point is that jpeg is lossy like some tiff compression. So why get rid of the compression in one case, but apply it in the other? IM could apply minimal compression to a jpeg. 

But ok my question got answered. I’ll throw up a feature request anyway. 

Finally it would be great if the docs stated this somewhere. IIRC there’s just the maintain compression flag but not a clear indication that normally it won’t be.
